File: backend/internal/statement/formula.go
package statement

import (
	"bytes"
	"strings"

	"app/internal/expression"
	fn "app/internal/functional"
	"app/internal/token"
)

type Formula struct {
	name   token.Token
	params []expression.Expression
}

func NewFormula(name token.Token, params []expression.Expression) *Formula {
	return &Formula{
		name:   name,
		params: params,
	}
}

func (formula *Formula) ToString() string {
	exprStrings := fn.Map(formula.params, exprToString)
	exprJoins := strings.Join(exprStrings, ", ")

	var buffer bytes.Buffer
	buffer.WriteString(formula.name.Literal)
	buffer.WriteString(" ")
	buffer.WriteString(exprJoins)
	return buffer.String()
}

func exprToString(expr expression.Expression) string {
	return expr.PrettyString()
}

File: backend/internal/statements/formula.py
from backend.internal.expressions import Expression
from backend.internal.statements import Statement
from backend.internal.tokens import Token


class Formula(Statement):
    def __init__(self, name: Token, params: list[Expression]) -> None:
        self._name = name
        self._params = params

    def __repr__(self) -> str:
        return f"FORMULA({str(self)})"

    def __str__(self) -> str:
        params_str = ", ".join(map(str, self._params))
        return f"'{self._name.literal}' {params_str}"

    def __eq__(self, other) -> bool:
        return (
            isinstance(other, Formula)
            and self._name == other._name
            and self.params == other.params
        )

    @property
    def name(self) -> Token:
        return self._name

    @property
    def params(self) -> list[Expression]:
        return self._params
